infra_event_notifier/backends/jira.py

```diff
@@ -1,8 +1,240 @@
+        self.tags = tags
+        self.issue_type = issue_type
+
+
+class JiraApiException(Exception):
+    def __init__(self, stderr: str):
+        self.message = stderr
+        super().__init__(self.message)
+
+
+def http_basic_auth(username: str, api_key: str, req: Request) -> Request:
+    """
+    Does the equivalent of requests' HTTPBasicAuth but for a urllib Request
+
+    Args:
+        username (str): Username for basic auth
+        api_key (str): API key for basic auth
+        req (Request): Request object to attach headers to
+    """
+    base64string = b64encode(bytes(f"{username}:{api_key}", "utf-8"))
+    req.add_header("Authorization", f"Basic {base64string.decode('utf-8')}")
+    return req
+
+
+def create_or_update_issue(
+    jira: JiraConfig,
+    fields: JiraFields,
+    fallback_comment_text: str | None,
+    update_text_body: bool = False,
 ) -> None:
-    # TODO: implement
-    pass
+    """
+    Attempts to create a Jira issue with the given title/text/tags.
+    If an issue matching that title and tags already exists, optionally update
+    that issue's text body and add a comment.
+
+    Args:
+        jira (JiraConfig): Config containing URL/project/email/API Key
+        fields (JiraFields): Fields of the Jira Issue
+        fallback_comment_text (str | None, optional): Optional comment to
+            include on jira issue if issue already exists. Defaults to None.
+        update_text_body (bool, optional): If set, will update the
+            body of an existing Jira issue with whatever is passed
+            as the `text` parameter. Defaults to False.
+    """
+    key = _find_jira_issue(jira, fields.title, fields.tags)
+    if key is not None:
+        if update_text_body:
+            _update_jira_issue(jira, key, fields.text)
+        if fallback_comment_text:
+            _add_jira_comment(jira, key, fallback_comment_text)
+    else:
+        _create_jira_issue(
+            jira, fields.title, fields.text, fields.tags, fields.issue_type
+        )
+
+
+def _create_jira_issue(
+    jira: JiraConfig,
+    title: str,
+    body: str,
+    tags: Mapping[str, str],
+    issue_type: str,
+) -> Any | None:
+    """
+    Attempts to create a new jira issue.
+    """
+    api_key = jira.api_key
+    api_url = f"{jira.url}/rest/api/2/issue"
+    payload = {
+        "fields": {
+            "project": {"key": jira.project_key},
+            "summary": title,
+            "description": body,
+            "issuetype": {"name": issue_type},
+            "labels": [f"{k}:{v}" for k, v in tags.items()],
+        }
+    }
+    json_data = json.dumps(payload)
+    data = json_data.encode("utf-8")
+    req = Request(api_url, data=data, method="POST")
+    req = http_basic_auth(jira.user_email, api_key, req)
+    req.add_header("Content-Type", "application/json")
+
+    with urllib.request.urlopen(req) as response:
+        status = response.status
+        if status == 201:
+            return response
```

Comment:
Why do you need to pass the response back. You never use it. If you decide to pass it back, please provide a type rather than Any